src/wp-includes/class-wp-theme.php

```diff
@@ -1476,17 +1484,49 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
 	 * @return bool
 	 */
 	public function is_block_theme() {
+		$is_block_theme = $this->cache_get( 'is_block_theme' );
+
+		// Use the cache, if set.
+		if ( false !== $is_block_theme ) {
+			return (bool) $is_block_theme;
+		}
+
+		// Short-circuit when the 'full-site-editing' tag is found.
+		if ( ! empty( $this->headers['Tags'] ) && str_contains( $this->headers['Tags'], 'full-site-editing' ) ) {
+			$this->cache_add( 'is_block_theme', 1 );
+			return true;
+		}
+
 		$paths_to_index_block_template = array(
 			$this->get_file_path( '/block-templates/index.html' ),
 			$this->get_file_path( '/templates/index.html' ),
 		);
 
 		foreach ( $paths_to_index_block_template as $path_to_index_block_template ) {
-			if ( is_file( $path_to_index_block_template ) && is_readable( $path_to_index_block_template ) ) {
+			if ( ! is_file( $path_to_index_block_template ) || ! is_readable( $path_to_index_block_template ) ) {
+				continue;
+			}
+
+			$content = file_get_contents( $path_to_index_block_template );
+
+			// Skip a file that has no blocks.
+			if ( ! str_contains( $content, '<!-- wp:' ) ) {
+				continue;
+			}
+
+			$blocks = parse_blocks( trim( $content ) );
```

Comment:
Without `trim()`, `parse_blocks()` will interpret an empty last line as a block.

Reply:
**questions (non-blocking):** Is there a reason we're parsing the blocks? The `str_contains` should be enough to check the existence of the blocks.

The `has_blocks` uses a similar check as well - https://developer.wordpress.org/reference/functions/has_blocks/.

Reply:
Hybrid themes may use mixed HTML (block and non-block markup). The thinking here is that parsing the blocks is a way to determine whether this is a "pure" block template or not.

The `str_contains()` is a quick way for us to determine if there are _any_ blocks. If there are no blocks, we can return `false` and save the resources required to parse the contents more thoroughly.